samza-core/src/main/java/org/apache/samza/table/batching/AsyncBatchingTable.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table.batching;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import org.apache.samza.SamzaException;
+import org.apache.samza.context.Context;
+import org.apache.samza.storage.kv.Entry;
+import org.apache.samza.table.AsyncReadWriteTable;
+import org.apache.samza.table.ReadWriteTable;
+import org.apache.samza.table.utils.TableMetricsUtil;
+
+
+/**
+ * A wrapper of a {@link AsyncReadWriteTable} that supports batch operations.
+ *
+ * This batching table does not guarantee any ordering of different operation types within the batch.
+ * For instance, query(Q) and update(u) operations arrives in the following sequences, Q1, U1, Q2, U2,
+ * it does not mean the the remote data store will receive the messages in the same order. Instead,
+ * the operations will be grouped by type and sent via micro batches. For this sequence, Q1 and Q2 will
+ * be grouped to micro batch B1; U1 and U2 will be grouped to micro batch B2, the implementation class
+ * can decide the order of the micro batches.
+ *
+ * Synchronized table operations (get/put/delete) should be used with caution for the batching feature.
+ * If the table is used by a single thread, there will be at most one operation in the batch, and the
+ * batch will be performed when the TTL of the batch window expires. Batching does not make sense in this scenario.
+ *
+ * @param <K> The type of the key.
+ * @param <V> The type of the value.
+ */
+public class AsyncBatchingTable<K, V> implements ReadWriteTable<K, V> {
```

Comment:
For delegation tables, we only need to implement the async methods. So it only need to implement AsyncReadWriteTable